example/audio/1-oscillator.cpp

```diff
@@ -0,0 +1,150 @@
+
+namespace {
+using namespace mp_units;
+
+//! A DSP generator class that generates sample values for a sine wave oscillator.
+class sine_wave_osc {
+public:
+  sine_wave_osc(const audio::musical_context& context, QuantityOf<isq::frequency> auto freq) :
+      m_context{context}, m_frequency{freq}
+  {
+    std::cout << MP_UNITS_STD_FMT::format(
+      "Created oscillator with starting frequency {} ({}) for sample rate {} at tempo {}\n", freq, m_frequency,
+      context.current_sample_rate, context.current_tempo);
+  }
+
+  quantity<si::hertz, float> get_frequency() const { return m_frequency; }
+
+  void set_frequency(QuantityOf<isq::frequency> auto freq)
+  {
+    m_frequency = freq;
+    std::cout << MP_UNITS_STD_FMT::format("Setting frequency to {} ({})\n", freq, m_frequency);
+  }
+
+  void set_period(QuantityOf<isq::time> auto period)
+  {
+    m_frequency = 1.f / value_cast<float>(period);
+    std::cout << MP_UNITS_STD_FMT::format("Setting period to {} (i.e. frequency to {})\n", period, m_frequency);
+  }
+
+  void set_period(QuantityOf<audio::beat_count> auto period)
+  {
+    std::cout << MP_UNITS_STD_FMT::format("Setting period to {} -- ", period);
+    set_period(value_cast<float>(period) / m_context.current_tempo);
+  }
+
+  quantity<audio::sample_value, float> operator()()
+  {
+    auto out = angular::sin(m_phase.quantity_from_zero());
+    m_phase += m_step;
+    return out;
+  }
+
+  void reset() { m_phase = phase_t{0.f * angular::radian}; }
+
+private:
+  using phase_t = quantity_point<angular::radian, mp_units::default_point_origin(angular::radian), float>;
+
+  void update_step() { m_step = (m_frequency / m_context.current_sample_rate) * angular::revolution; }
+
+  audio::musical_context m_context;
+  quantity<si::hertz, float> m_frequency;
+  phase_t m_phase{0.f * angular::radian};
+  quantity<angular::radian, float> m_step;
+};
+}  // namespace
+
+int main()
+{
+  using namespace mp_units::si::unit_symbols;
+
+  const auto context = audio::get_musical_context();
+
+  // Sine oscillators are sometimes used as a "low-frequency oscillator"
+  // (LFO) that runs at a frequency below the range of human hearing and
+  // is used a source of modulation for other paramters in an audio
+  // algorithm.
+  auto sin_gen = sine_wave_osc{context, 1 * Hz};
+
+  // Depending on the use-case sometimes an LFO will be set with a frequency in Hz
+  sin_gen.set_frequency(13 * Hz);
+
+  // for some use-cases it is more convient for a user to set the period
+  sin_gen.set_period(42 * s);
+
+  // and in some other use-cases setting the period in musical note duration is more intuitive
+  sin_gen.set_period(1 * audio::half_note);
+
+  // Our oscillator can be used to generate sample values for a buffer
+  // of audio samples.  In this example we will create a buffer with
+  // duration equal to 2 measures of 4/4 music (i.e. 2 whole notes at
+  // the current tempo):
+  const auto beats = 2 * audio::whole_note;
```

Comment:
It is up to you, and this is purely a matter of personal preference. :wink: AAA (Almost Always Auto) is a nice style, and I like it in many cases (especially not to repeat myself).

However, with time, I have learned to appreciate CTAD as it makes the code easier to reason about. Using `quantity` and `quantity_point` CTAD placeholders instead of `auto` probably makes the code easier for a new user to understand or maintain.

Reply:
`const Quantity auto beats` is also an option, but it is more to type and does not have any benefits here over just `const quantity beats`.